File: Sources/MoonRuntime/DeepSeekApi.swift
import Foundation

public enum DeepSeekApiFormat: String, Sendable {
    case openai
    case anthropic
}

public struct ResolvedDeepSeekApi: Sendable {
    public var format: DeepSeekApiFormat
    public var baseUrl: String
    public var endpoint: String
    public var useBeta: Bool
}

public let deepSeekOpenAiBase = "https://api.deepseek.com"
public let deepSeekAnthropicBase = "https://api.deepseek.com/anthropic"
public let deepSeekBetaBase = "https://api.deepseek.com/beta"

public func detectApiFormat(_ baseUrl: String) -> DeepSeekApiFormat {
    baseUrl.contains("/anthropic") ? .anthropic : .openai
}

public func normalizeBaseUrl(_ baseUrl: String) -> String {
    var url = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    if url.hasSuffix("/v1") {
        url = String(url.dropLast(3))
    }
    return url
}

public func resolveDeepSeekApi(
    baseUrl: String? = nil,
    apiFormat: DeepSeekApiFormat? = nil,
    useBeta: Bool? = nil
) -> ResolvedDeepSeekApi {
    let envBeta = ProcessInfo.processInfo.environment["DEEPSEEK_USE_BETA"] == "true"
        || ProcessInfo.processInfo.environment["DEEPSEEK_USE_BETA"] == "1"

    let envFormatRaw = ProcessInfo.processInfo.environment["DEEPSEEK_API_FORMAT"]?.lowercased()
    let formatFromEnv: DeepSeekApiFormat? = envFormatRaw == "anthropic" ? .anthropic
        : envFormatRaw == "openai" ? .openai
        : nil

    var resolvedBase = normalizeBaseUrl(
        baseUrl
            ?? ProcessInfo.processInfo.environment["DEEPSEEK_BASE_URL"]
            ?? deepSeekAnthropicBase
    )

    var format = apiFormat ?? formatFromEnv ?? detectApiFormat(resolvedBase)
    let beta = useBeta ?? envBeta

    if beta && format == .openai {
        resolvedBase = deepSeekBetaBase
    }

    if format == .anthropic && !resolvedBase.contains("/anthropic") {
        resolvedBase = deepSeekAnthropicBase
    }

    let endpoint = format == .anthropic ? "/v1/messages" : "/chat/completions"
    return ResolvedDeepSeekApi(format: format, baseUrl: resolvedBase, endpoint: endpoint, useBeta: beta && format == .openai)
}

File: legacy/packages/runtime/src/deepseek-api.ts
export type DeepSeekApiFormat = "openai" | "anthropic";

export interface ResolvedDeepSeekApi {
  format: DeepSeekApiFormat;
  baseUrl: string;
  endpoint: string;
  useBeta: boolean;
}

export const DEEPSEEK_OPENAI_BASE = "https://api.deepseek.com";
export const DEEPSEEK_ANTHROPIC_BASE = "https://api.deepseek.com/anthropic";
export const DEEPSEEK_BETA_BASE = "https://api.deepseek.com/beta";

export function detectApiFormat(baseUrl: string): DeepSeekApiFormat {
  if (baseUrl.includes("/anthropic")) return "anthropic";
  return "openai";
}

export function normalizeBaseUrl(baseUrl: string): string {
  return baseUrl.replace(/\/+$/, "").replace(/\/v1$/, "");
}

export function resolveDeepSeekApi(options: {
  baseUrl?: string;
  apiFormat?: DeepSeekApiFormat;
  useBeta?: boolean;
}): ResolvedDeepSeekApi {
  const envBeta = process.env.DEEPSEEK_USE_BETA === "true" || process.env.DEEPSEEK_USE_BETA === "1";
  const useBeta = options.useBeta ?? envBeta;

  const envFormat = process.env.DEEPSEEK_API_FORMAT?.toLowerCase();
  const formatFromEnv = envFormat === "anthropic" || envFormat === "openai"
    ? envFormat
    : undefined;

  let baseUrl = normalizeBaseUrl(
    options.baseUrl
    ?? process.env.DEEPSEEK_BASE_URL
    ?? DEEPSEEK_ANTHROPIC_BASE,
  );

  let format = options.apiFormat ?? formatFromEnv ?? detectApiFormat(baseUrl);

  if (useBeta && format === "openai") {
    baseUrl = DEEPSEEK_BETA_BASE;
  }

  if (format === "anthropic" && !baseUrl.includes("/anthropic")) {
    baseUrl = DEEPSEEK_ANTHROPIC_BASE;
  }

  const endpoint = format === "anthropic" ? "/v1/messages" : "/chat/completions";

  return { format, baseUrl, endpoint, useBeta: useBeta && format === "openai" };
}